crates/solidity/testing/snapshots/bindings_assertions/interfaces/visibility.sol

```diff
@@ -22,9 +22,11 @@ contract Test {
     //<ref:1
     //      ^ref:3
 
-    function my_test() returns (int) {
-        return Example.calculate();
-        //             ^ref:!
-        //     ^ref:1
+    function my_test(Example e) returns (int) {
```

Comment:
Just for my own understanding, I wonder how would a test like this benefit from being in `bindings_assertions` instead of `bindings_output`?

I don't see any special assertions, and they can all be updated automatically like the others..

Reply:
I've been having second thoughts about binding assertions myself. On the one hand, it's useful to have a loud failing test whenever I make a change that breaks the binding semantics. On the other, it usually gets a bit cumbersome to debug those failures because the assertions don't generate a graph output by default, which is my go-to tool to reason about rules output. So I ended up making an external tool to generate the `.dot` file from any given source, similar to what we get for `bindings_output` but manual.

Also, the assertions are a bit behind in feature parity with outputs, which is something I've been putting off updating for a while: they don't support asserting built-in definitions nor ambiguous results.

As a result, I seldom write any new assertion cases.

Reply:
Almost all of the tests in this repo self-update, which is really helpful in large-scale changes that we often do. So having to manually fix/update tests can be time-consuming.

If the main benefit is having a "loud failing test", I wonder if I can help fixing this issue, by making the snapshot tests more prominent/easier to read somehow?

We can think about this later of course.

Reply:
Hmmm... probably this *can* be improved with some tooling as you suggest. 

TOL: Currently, my workflow when changing the binding rules is, roughly:
1. Write a new case that fails to resolve for the bug I'm fixing
2. Fix the rules
3. Run test for the single new case
4. Repeat 2 and 3 until the new case passes
5. Run all the tests

At this point, if I broke something either some assertion fails with hopefully a more or less clear message of what's wrong, or some of the outputs of the `bindings_output` cases changed. For the second thing I need to actively look for changes and then analyzing what changed can be a bit confusing. Usually some `-success.txt` file got deleted and a new `-failure.txt` file is created in its place. Other times there are changes in failure files, if we're dealing with version specific changes. That inspection is all manual anyway. Best case scenario none of the snapshots changed, meaning nothing broke.

It should be possible to make the diff inspection more automated.